### PR TITLE
8301489: C1: ShortLoopOptimizer might lift instructions before their inputs

### PR DESCRIPTION
**Pull request by danielogh**
ShortLoopOptimizer might lift instructions before their inputs on some graph shapes. We propose adding a check that the insertion point for an instruction that is a candidate for hoisting should not be higher up the dominator tree than any inputs to the instruction.

Testing: tier1-tier3.

Additional testing: observed that `(cur_invariant && !v.is_valid())` never occurs on tier1-tier3 before the added test case.
Also verified that the depth check is equivalent to `(*vp->block() == _insert->block()) || dominates(*vp, _insert)` on all of tier1-tier3.

Failure case: in the attached image the `arraylength` instruction from B10 is lifted to B0, as the dominator of B10 is calculated as B0. This is based on the logic in [`ComputeLinearScanOrder::compute_dominator_impl`](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/c1/c1_IR.cpp#L801).  But the array input is in Block 3. This is later spotted in `c1_LIRAssembler.cpp` with `Error: ShouldNotReachHere()`. We can reproduce the error on other instructions too -- the reader may refer to the test case provided.

image

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301489](https://bugs.openjdk.org/browse/JDK-8301489): C1: ShortLoopOptimizer might lift instructions before their inputs (**Bug** - P3)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14492/head:pull/14492` \
`$ git checkout pull/14492`

Update a local copy of the PR: \
`$ git checkout pull/14492` \
`$ git pull https://git.openjdk.org/jdk.git pull/14492/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14492`

View PR using the GUI difftool: \
`$ git pr show -t 14492`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14492.diff">https://git.openjdk.org/jdk/pull/14492.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14492#issuecomment-1592857723)

### CODE REVIEW

**Review by TobiHartmann — COMMENTED** [review]
Great work investigating this, Daniel! 

From your comments in JBS, it seems that the underlying issue is additional exception edges in the graph that affect dominator computation. Could you elaborate a bit more on that with respect to the example that you provided in the PR description?

I'm not an expert in C1 though (paging @veresov and @rwestrel as the author of JDK-7153771).

Thanks,
Tobias

File: src/hotspot/share/c1/c1_ValueMap.cpp [hunk 1]
```diff
@@ -359,6 +359,28 @@ LoopInvariantCodeMotion::LoopInvariantCodeMotion(ShortLoopOptimizer *slo, Global
   }
 }
 
+class CheckInsertionPoint: public ValueVisitor {
```
Comment (robcasloz) [inline src/hotspot/share/c1/c1_ValueMap.cpp:362]: For consistency with other similar cases in `c1_ValueMap.cpp`: 
```suggestion
class CheckInsertionPoint : public ValueVisitor {
```

File: src/hotspot/share/c1/c1_ValueMap.cpp [hunk 2]
```diff
@@ -394,7 +416,11 @@ void LoopInvariantCodeMotion::process_block(BlockBegin* block) {
       cur_invariant = is_invariant(cvt->value());
     }
 
-    if (cur_invariant) {
+    // Check that insertion point has higher dom depth than all inputs to cur
+    CheckInsertionPoint v(_insertion_point);
+    cur->input_values_do(&v);
```
Comment (robcasloz) [inline src/hotspot/share/c1/c1_ValueMap.cpp:421]: For compilation efficiency, would it be possible to perform this computation only when `cur_invariant` holds? You could for example encapsulate the creation of `v`, call to `cur->input_values_do(&v)`, and `v.is_valid()` check into an auxiliary function `is_dominated_by_inputs(_insertion_point, cur)` or similar and call that function below (`if (cur_invariant && is_dominated_by_inputs(_insertion_point, cur)) {`).

**Review by robcasloz — CHANGES_REQUESTED** [review]
Good catch, analysis, and regression test Daniel. Great that you could generalize the fix to cover all other LICM cases. Just a couple of comments/suggestions.

File: src/hotspot/share/c1/c1_ValueMap.cpp [hunk 3]
```diff
@@ -359,6 +359,28 @@ LoopInvariantCodeMotion::LoopInvariantCodeMotion(ShortLoopOptimizer *slo, Global
   }
 }
 
+class CheckInsertionPoint: public ValueVisitor {
+ private:
+  Value _insert;
+  bool _valid = true;
+
+  void visit(Value* vp) {
```
[the remark this answers is not in the document]
Comment (danielogh) [reply src/hotspot/share/c1/c1_ValueMap.cpp:367]: I am not sure if this is possible without changing the ValueVisitor ([ref](https://java.se.oracle.com/source/xref/jdk-jdk/jdk-open/src/hotspot/share/c1/c1_Instruction.hpp#123)) itself

**Comment by danielogh** [comment]
> Great work investigating this, Daniel!
> 
> From your comments in JBS, it seems that the underlying issue is additional exception edges in the graph that affect dominator computation. Could you elaborate a bit more on that with respect to the example that you provided in the PR description?
> 
> I'm not an expert in C1 though (paging @veresov and @rwestrel as the author of JDK-7153771).
> 
> Thanks, Tobias

Thank you very much for review! 

My understanding is, this might have been introduced in JDK-7153771, where we have extra edges to the exception handler of all successors during dominator calculation -- `Additional edge to xhandler of all our successors`. The client RangeCheckElimination optimization needed this additional information, but the short loop optimization was not updated accordingly. In this specific case, I observed using print debugging and -XX:TraceLinearScanLevel=4, that the dominator of B10 is being calculated as common_dominator(B7, B1). But B1 has dominator B0 because of the extra edge from B14 to B1 (the exception handler of B3). The short loop optimization lifts loop invariant instructions to the dominator of the loop header, which in this case becomes B0.

**Review by robcasloz — APPROVED** [review]
Thanks for addressing the suggestions and for the additional explanation, looks good!

File: src/hotspot/share/c1/c1_ValueMap.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (TobiHartmann) [reply src/hotspot/share/c1/c1_ValueMap.cpp:367]: Right, makes sense.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by danielogh** [comment]
Thanks Roberto and Tobias for review. Thanks Roberto for additional help with PR.

**Pull request closed without merging**